File: linear_model.py
import sys
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

if len(sys.argv) != 4:
    print("Usage: python linear_regression_python.py <filename> <x_column> <y_column>")
    sys.exit(1)

filename = sys.argv[1]
x_col = sys.argv[2]
y_col = sys.argv[3]

data = pd.read_csv("regression_data.csv")
model = LinearRegression()
model.fit(data[[x_col]], data[[y_col]])

plt.scatter(data[[x_col]], data[[y_col]], color='red')
plt.plot(data[[x_col]], model.predict(data[[x_col]]), color='blue')
plt.title(f'{y_col} vs {x_col}')
plt.xlabel(x_col)
plt.ylabel(y_col)
plt.savefig("regression_plot_python.png")
plt.show()

#!/usr/bin/env python
# coding: utf-8

# This notebook demonstrates a simple linear regression analysis using Python to model Salary based on Years of Experience.

# In[1]:


pip install pandas


# In[2]:


import pandas as pd # Installing Pandas to allow regression to run


# In[3]:


dataset = pd.read_csv("regression_data.csv") # Letting Py use the csv as "dataset"


# In[4]:


pip install matplotlib


# In[5]:


import matplotlib.pyplot as plt # Installing matplotlib


# In[6]:


plt.scatter(dataset["YearsExperience"], dataset["Salary"], color="red")
# Creating our initial scatterplot


# In[10]:


pip install scikit-learn


# In[11]:


from sklearn.linear_model import LinearRegression # Importing the linearregression function to run the regression
reg = LinearRegression() # Naming the linear regression as "reg"
reg.fit(dataset[["YearsExperience"]], dataset[["Salary"]]) 


# In[13]:


plt.plot(dataset["YearsExperience"], reg.predict(dataset[["YearsExperience"]]), color="blue")
plt.title("Salary vs Experience")
plt.xlabel("Years of Experience")
plt.ylabel("Salary")
plt.show()
# Plotting the regression line


# In[14]:


reg.score(dataset[["YearsExperience"]], dataset[["Salary"]])
# Displaying the R-squared value for our regression


# In[16]:


import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error

x = np.array(dataset.YearsExperience)
y = np.array(dataset.Salary)

slope, intercept, r_value, p_value, std_err = linregress(x, y)
y_pred = slope * x + intercept
mse = mean_squared_error(y, y_pred)

plt.plot(x, y_pred, 'r-', label='Fitted Line')
plt.text(1.5, max(y) - 1,
         f"y = {slope:.2f}x + {intercept:.2f}\n"
         f"r = {r_value:.2f}\nMSE = {mse:.2f}",
         fontsize=12)
plt.xlabel("x")
plt.ylabel("y")
plt.title("Regression")
plt.legend()
plt.savefig("regression_plot_python.png") #saving the png file of the regression
plt.show()


# In[17]:


print(slope) # Slope
print("This is the slope for our regression model")


# In[18]:


print(intercept)
print("This is the intercept for our regression model")


# In[19]:


print(r_value)
print("This is the r-value for our regression model")


# In[20]:


print(mse)
print("This is the MSE for our regression model")


# In[ ]:

File: linear_model.r
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  stop("Usage: Rscript linear_model.r <filename> <x_column> <y_column>")
}

filename <- args[1]
x_col <- args[2]
y_col <- args[3]

data <- read.csv("regression_data.csv")
formula <- as.formula(paste(y_col, "~", x_col))
model <- lm(formula, data = data)

library(ggplot2)
plot <- ggplot(data, aes_string(x = x_col, y = y_col)) +
  geom_point(color = "red") +
  geom_smooth(method = "lm", color = "blue") +
  ggtitle(paste(y_col, "vs", x_col)) +
  xlab(x_col) +
  ylab(y_col)

ggsave("linear_regression_r_output.png", plot)
print(plot)

dataset <- read.csv("regression_data.csv") # Reading in the csv file

print(dataset) # Double Checking the dataset to make sure everything worked.

plot(dataset$YearsExperience, dataset$Salary, col="red")

model <- lm(Salary ~ YearsExperience, data=dataset) # Making a linear model of the dataset in order to run a regression analysis

library(ggplot2) # Installing ggplot2 in order to run the regression.

library(ggplot2)
ggplot() +
    geom_point(aes(x = dataset$YearsExperience, y = dataset$Salary), colour = 'blue') +
    geom_line(aes(x = dataset$YearsExperience, y = predict(model, newdata = dataset)), colour = 'black') +
    ggtitle('Salary vs Experience') +
    xlab('Years of Experience') +
    ylab('Salary')

summary(model)

library(ggplot2)

# Data
df <- data.frame(
  x = c(dataset$YearsExperience),
  y = c(dataset$Salary)
)

# Fit model
model <- lm(y ~ x, data = df)
slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(df$x, df$y)
pred <- predict(model)
mse <- mean((df$y - pred)^2)

# Plot
ggplot(df, aes(x = x, y = y)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "red") +
  annotate("text", x = 1.5, y = max(df$y) - 0.5,
           label = paste("y =", round(slope, 2), "x +", round(intercept, 2),
                         "\nr =", round(r, 2), "\nMSE =", round(mse, 2)),
           size = 4) +
  labs(title = "Linear Fit",
       x = "x", y = "y") +
  theme_minimal()

ggsave("regression_plot_r.png")


